### DISCUSSION TITLE
how to render under tabs

### DISCUSSION THREAD

**Discussion by guozhi72** [Kitware/trame#269] [Q&A] [answered]
Hi, 
I try to have several tabs. Once click tab A, render cone, click tab B, render a 3D volume, for example head.vti. Try to search, but totoally have no idea how to do it.

If I look through vuetify Tabs, I can easisy find how to handle the tab. But with trame, just have no idea at all.

Can someone show me how to do this? Specially, how to relate vuetify with trame.vuetify. I really have trouble to handle trame.vuetify.
<template>
  <v-card>
    <v-tabs
      v-model="tab"
      bg-color="deep-purple-accent-4"
      centered
      stacked
    >
      <v-tab value="tab-1">
        <v-icon>mdi-phone</v-icon>
        Recents
      </v-tab>

      <v-tab value="tab-2">
        <v-icon>mdi-heart</v-icon>
        Favorites
      </v-tab>

      <v-tab value="tab-3">
        <v-icon>mdi-account-box</v-icon>
        Nearby
      </v-tab>
    </v-tabs>

    <v-window v-model="tab">
      <v-window-item
        v-for="i in 3"
        :key="i"
        :value="'tab-' + i"
      >
        <v-card>
          <v-card-text>{{ text }}</v-card-text>
        </v-card>
      </v-window-item>
    </v-window>
  </v-card>
</template>

**Comment by jourdain** [comment]
```python
with layout:
   with vuetify.VTabs():
      with vuetify.VTab(value="tab-2") as tab:
         vuetify.VIcon("mdi-heart")
         tab.add_child("Favorites")
      with vuetify.VTab(value="tab-3") as tab:
         vuetify.VIcon("mdi-account-box")
         tab.add_child("Nearby")

    with vuetify.VWindow(v_model="tab"):
       with vuetify.VWindowItem(
           v_for="i in 3",
           key="i",
           value=("`tab-${i}`",),
       ):
           with vuetify.VCard():
               vuetify.VCardText("{{ text }}")

```

**Comment by jourdain** [comment]
You can double check by calling `print(layout.html)`. For more information refer to course reference material [here](https://kitware.github.io/trame/docs/course_intro.html).

**Comment by guozhi72** [comment]
Hi, Jourdain:
Thanks for immediate response. I am not able to make it work. But looking your code, I got the idea what to do. However, for my below code, the server has stop responding. The idea what I see is if value = 'tabl-0', will render the cone, otherwise render else.
    def layout(self):
        # with RouterViewLayout(self.server, "/"):
        with vuetify.VCard():
            vuetify.VCardTitle("This is home")
            vuetify.VBtn(
                "my tabs page",
                click=self.send_to_server1
            )

        with vuetify.VTabs(v_model=("tab", 'tab-0')):
            with vuetify.VTab(value="tab-0") as tab_1:
                vuetify.VIcon("mdi-heart")
                tab_1.add_child("Favorites")
            with vuetify.VTab(value="tab-1") as tab_2:
                vuetify.VIcon("mdi-account-box")
                tab_2.add_child("Nearby")

        with vuetify.VWindow(v_model="tab"):
            with vuetify.VWindowItem(value="tab-0"):
                #  rener cone here
                vuetify.VBtn(
                    "tab 0",)
            with vuetify.VWindowItem( value="tab-1") as w2:
                #  render head volume here
                vuetify.VBtn(
                    "tab 1",)

**Reply by jourdain** [reply]
Please post the full code otherwise we can only give you the gist of it not a correct and valid code.

**Reply by jourdain** [reply]
Also it seems that you are aiming to use vuetify3 but in Python you are using vuetify2? Is that correct?

**Comment by guozhi72** [comment]
I just start using vue, not sure whether I am using vue2 or vue3. Before I really to use vue, I am trying to set up a little more complex systemfor example:

1 mutiple tabs, click each tab will rendering different data.
2, mutiple pages rendering with routing.
3, it is a pity I can not seperate ui with rendering using trame, but at least I want to try more complex layout first.

Attached is my entire code so far, I try to
1, routers, successed.
2, code sepreation, so far, look ok to me.
3, tabs, if I click, server will stop responding.
4, I try to redning cone and head in same volume page, but the two rending are either same, unless I use localView for one, remoteView for anotehr.
5, samething for /cone and /volume routes, I am expecting to see cone in /cone page, cone and voluem in /volume route. But I will always see same thing.



    On Monday, June 19, 2023 at 05:37:26 PM CDT, Sebastien Jourdain ***@***.***> wrote:  
 
 


Also it seems that you are aiming to use vuetify3 but in Python you are using vuetify2? Is that correct?

**Comment by guozhi72** [comment]
I just start using trame, not sure whether I am using vue2 or vue3. Before I really to use trame,  I am trying to set up a little more complex system( sorry for typo in my reply)

**Reply by jourdain** [reply]
Right now the Python by default is using vue2. So make sure when looking at the code online, you refer to vuetify for vue 2.

**Comment by jourdain** [comment]
Still don't see the full code but I when you use several 3d view, make sure each of the view have a unique `ref="viewCone"` vs `ref="viewVolume"`.

**Reply by guozhi72** [reply]
[rendering_trame - Copy.zip](https://github.com/Kitware/trame/files/11802201/rendering_trame.-.Copy.zip)
I attached the entire zip. Let me can you see the source code, now mainly want to solve two things,
1, how to do rendering under tab, now if click the tap, the server freeze.
2, how to do conditional rendering within trame frame work.

**Comment by jourdain** [comment]
Solved on some duplicated thread

**Reply by guozhi72** [reply]
I could not make the tabs works. When I click the tab, server freeze. Entire code is posted before.

**Reply by jourdain** [reply]
I don't see the code in that discussion thread. (Use the web UI not your emails)

**Comment by guozhi72** [comment]
[rendering_trame - Copy.zip](https://github.com/Kitware/trame/files/11819106/rendering_trame.-.Copy.zip)
I attached the entire code through web ui, can you see it.  
1, major problem of this code if click the tab, server freeze

**Comment by jourdain** [comment, marked as answer]
```python
from trame.app import get_server
from trame.ui.vuetify import SinglePageLayout
from trame.widgets import html, vuetify, vtk as vtk_widgets

from vtkmodules.vtkFiltersSources import vtkConeSource
from vtkmodules.vtkRenderingCore import (
    vtkActor,
    vtkPolyDataMapper,
    vtkRenderer,
    vtkRenderWindow,
)

import vtkmodules.vtkRenderingOpenGL2  # noqa

server = get_server()

class Cone:
    def __init__(self, server, name):
        self.server = server
        self.name = name

        self.rw = vtkRenderWindow()
        renderer = vtkRenderer()
        self.rw.AddRenderer(renderer)
        cone_source = vtkConeSource()
        mapper = vtkPolyDataMapper()
        mapper.SetInputConnection(cone_source.GetOutputPort())
        actor = vtkActor()
        actor.SetMapper(mapper)
        renderer.AddActor(actor)
        renderer.ResetCamera()

        self.ui()

    @property
    def state(self):
        return self.server.state

    @property
    def ctrl(self):
        return self.server.controller

    def ui(self, **kwargs):
        with vtk_widgets.VtkLocalView(self.rw, ref=f"view_{self.name}") as view:
            html.Div(
                self.name,
                style="position: absolute; left: 10px; top: 10px; z-index: 100; color: white;",
            )
            self.ctrl.view_update = view.update
            self.ctrl.view_reset_camera = view.reset_camera


# ---------------------------------------------------------
# Main UI
# ---------------------------------------------------------

with SinglePageLayout(server) as layout:
    with layout.toolbar:
        layout.title.set_text("Active Tab: {{ active_tab }}")
        layout.title.style = "min-width: 200px;"
        vuetify.VSpacer()
        with vuetify.VTabs(v_model=("active_tab", 0), right=True):
            for i in range(5):
                vuetify.VTab(f"Cone {i + 1}")

    with layout.content:
        with vuetify.VContainer(classes="fill-height", fluid=True):
            with vuetify.VTabsItems(
                value=("active_tab",), style="width: 100%; height: 100%;"
            ):
                for i in range(5):
                    with vuetify.VTabItem(
                        value=(i,), style="width: 100%; height: 100%;"
                    ):
                        # html.Div(f"Instance {i + 1}", style="background: red; width: 100%; height: 100%;")
                        Cone(server, f"cone_{i + 1}")


if __name__ == "__main__":
    server.start()

```

**Comment by guozhi72** [comment]
thanks a lot.
